posts/2022/10-11-graphql_subscription_setup.md

```diff
@@ -0,0 +1,246 @@
+But to answer the question as to why am I writing this post: for future reference! Everything I know about subscriptions in one place. So let's get started!
+
+
+# 2. About the underlying app
+
+Before we dive into the subscription setup, I will describe what kind of app we are working with.
+
+Let's assume we have a Phoenix app with a GraphQL/Absinthe backend. We are using [GraphiQL](https://hexdocs.pm/absinthe_plug/Absinthe.Plug.GraphiQL.html), so we can play around with our queries, mutations and subscriptions.
+
+Our app has users, and we are using GraphQL queries to query users based on certain criteria and mutations to do things such as create or update users.
+All the usual stuff.
+
+The app is organized like this:
+
+```
+my_app_web   
+├── resolvers
+│   ├── some_resolver.ex 
+│   └── user.ex
+├── router.ex
+├── schema
+│   ├── mutations
+│   │   ├── some_mutation.ex
+│   │   └── user.ex
+│   ├── queries
+│   │   ├── some_query.ex
+│   │   └── user.ex
+│   └── subscriptions
+│       └── user.ex
+|── types
+|   ├── some_type.ex
+|   └── user.ex 
+├── schema.ex
+
+```
+
+We can see that the mutations, queries and types are defined in separate modules. I import them and their fields into the `schema.ex` file, which looks like this:
+
+```elixir
+defmodule MyAppWeb.Schema do
+  @moduledoc false
+  use Absinthe.Schema  
+
+  import_types MyAppWeb.Types.SomeType
+  import_types MyAppWeb.Types.User 
+  import_types GraphqlApiWeb.Schema.Queries.SomeQuery  
+  import_types GraphqlApiWeb.Schema.Queries.User  
+  import_types GraphqlApiWeb.Schema.Mutations.SomeMutation
+  import_types GraphqlApiWeb.Schema.Mutations.User 
+  import_types GraphqlApiWeb.Schema.Subscriptions.User
+
+  query do
+    import_fields :some_query_queries   
+    import_fields :user_queries   
+  end
+
+  mutation do
+    import_fields :some_mutations_mutations
+    import_fields :user_mutations
+  end
+
+  subscription do    
+    import_fields :user_subscriptions
+  end
+end
+```
+
+With this out of the way, let's get started!
+
+
+# 3. Subscription set-up - the infrastructure
+
+As in most cases, it is worth checking out the [hex.docs](https://hexdocs.pm/absinthe/subscriptions.html) first.
+Based on those and on what I have learned in the course, plus some personal experience, I have come up with the following steps for the basic setup:
+
+First, we need to add some dependencies (if we are running an app as described in 2. above, we'll already have those): 
+
+```elixir
+{:absinthe, "~> 1.6"},
+{:absinthe_plug, "~> 1.5"},
+{:absinthe_phoenix, "~> 2.0.0"}
+```
+
+Then we have to create a `user_socket.ex` file (it goes into the `my_app_web_folder`) and put this piece of code in it:
+
+```elixir
+defmodule MyAppWeb.UserSocket do
+  use Phoenix.Socket
+  use Absinthe.Phoenix.Socket,
+   schema: MyAppWeb.Schema
+ def connect(_params, socket) do
+  {:ok, socket}
+ end
+ def id(_socket), do: nil
+end     
```

Comment:
Want to put some spaces in here between things?

Reply:
I suppose :)